src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1387 @@
+    """
+    def __init__(self, config: ProphetNetConfig):
+        super().__init__()
+        self.embed_dim = config.hidden_size
+        self.self_attn = SelfAttention(
+            self.embed_dim, config.encoder_attention_heads, dropout=config.attention_dropout,
+        )
+        self.self_attn_layer_norm = LayerNorm(self.embed_dim)
+        self.dropout = config.dropout
+        self.activation_fn = ACT2FN[config.activation_function]
+        self.activation_dropout = config.activation_dropout
+        self.fc1 = nn.Linear(self.embed_dim, config.encoder_ffn_dim)
+        self.fc2 = nn.Linear(config.encoder_ffn_dim, self.embed_dim)
+        self.final_layer_norm = LayerNorm(self.embed_dim)
+
+    def forward(self, hidden_states, encoder_padding_mask, output_attentions=False):
+        residual = hidden_states
+        hidden_states, attn_weights = self.self_attn(
+            query=hidden_states, key=hidden_states, key_padding_mask=encoder_padding_mask, output_attentions=output_attentions
+        )
+        hidden_states = F.dropout(hidden_states, p=self.dropout, training=self.training)
+        hidden_states = residual + hidden_states
+        hidden_states = self.self_attn_layer_norm(hidden_states)
+        residual = hidden_states
+        hidden_states = self.fc1(hidden_states)
+        hidden_states = self.activation_fn(hidden_states)
+        hidden_states = F.dropout(hidden_states, p=self.activation_dropout, training=self.training)
+        hidden_states = self.fc2(hidden_states)
+        hidden_states = F.dropout(hidden_states, p=self.dropout, training=self.training)
+        hidden_states = residual + hidden_states
+        hidden_states = self.final_layer_norm(hidden_states)
+        return hidden_states, attn_weights
+
+
+
+class ProphetNetEncoder(nn.Module):
+    """
+    Same to Transformer Encoder.
+    Transformer encoder consisting of *config.encoder_layers* self attention layers. Each layer
+    is a :class:`EncoderLayer`.
+
+    Args:
+        config: ProphetNetConfig
+    """
+    def __init__(self, config: ProphetNetConfig, embed_tokens):
+        super().__init__()
+
+        self.output_hidden_states = config.output_hidden_states
+
+        self.dropout = config.dropout
+        embed_dim = embed_tokens.embedding_dim
+        self.padding_idx = embed_tokens.padding_idx
+        self.max_source_positions = config.max_position_embeddings
+
+        self.embed_tokens = embed_tokens
+        self.embed_scale = None
+        self.embed_positions = LearnedPositionalEmbedding(config.max_position_embeddings+1+self.padding_idx, embed_dim, self.padding_idx)
+
+        self.layers = nn.ModuleList([EncoderLayer(config) for _ in range(config.encoder_layers)])
+        self.emb_layer_norm = LayerNorm(embed_dim)
+
+    def forward(self, input_ids, attention_mask=None, output_attentions=False, return_dict=False):
+        # remove bos to be consistent with fairseq version
+        input_ids = input_ids[:, 1:]
```

Comment:
This should be removed. Don't like this "magic" here. It's very confusing for the user to always cut the first token -> just don't put it in then?! We'll see how we can handle this